shared/rSharp.cpp

```diff
@@ -27,53 +27,99 @@ typedef intptr_t(CORECLR_DELEGATE_CALLTYPE* CallFullTypeNameDelegate)(RSharpGene
 typedef int (CORECLR_DELEGATE_CALLTYPE* GetObjectDirectDelegate)(RSharpGenericValue* returnValue);
 typedef int (CORECLR_DELEGATE_CALLTYPE* CreateSEXPWrapperDelegate)(intptr_t pointer, RSharpGenericValue* returnValue);
 typedef int (CORECLR_DELEGATE_CALLTYPE* CreateInstanceDelegate)(const char*, RSharpGenericValue** objects, int num_objects, RSharpGenericValue* returnValue);
-typedef int (CORECLR_DELEGATE_CALLTYPE* LoadFromDelegate)(const char*, RSharpGenericValue *returnValue);
+typedef int (CORECLR_DELEGATE_CALLTYPE* LoadFromDelegate)(const char*, RSharpGenericValue* returnValue);
 
 char_t* MergeLibraryPath(const char_t* libraryPath, const char_t* additionalPath);
 void freeObject(RSharpGenericValue* instance);
 RSharpGenericValue* get_rSharp_generic_value_from_s4(SEXP clrObj);
 RSharpGenericValue createInstance(char* ns_qualified_typename, RSharpGenericValue** parameters, const R_len_t numberOfObjects);
 RSharpGenericValue getCurrentObjectDirect();
 
+/////////////////////////////////////////
+// Functions with R specific constructs, namely SEXPs
+/////////////////////////////////////////
+SEXP make_int_sexp(int n, int* values) {
+	SEXP result;
+	long i = 0;
+	int* int_ptr;
+	PROTECT(result = NEW_INTEGER(n));
+	int_ptr = INTEGER_POINTER(result);
+	for (i = 0; i < n; i++) {
+		int_ptr[i] = values[i];
+	}
+	UNPROTECT(1);
+	return result;
+}
+
 /////////////////////////////////////////
 // Initialization and disposal of the CLR
 /////////////////////////////////////////
-void rSharp_create_domain(char ** libraryPath)
+SEXP rSharp_create_domain(SEXP args)
 {
-	// if already loaded in this process, do not load again
-	if (load_assembly_and_get_function_pointer != nullptr)
-		return;
-
-	// STEP 1: Load HostFxr and get exported hosting functions
-	if (!load_hostfxr())
+	int returnValue = 0;
+	try
 	{
-		assert(false && "Failure: load_hostfxr()");
-		return;
-	}
+		args = CDR(args);
+
+		const SEXP element = CAR(args);
+		const char* libraryPath = CHAR(STRING_ELT(element, 0));
+
+		// if already loaded in this process, do not load again
+		if (load_assembly_and_get_function_pointer != nullptr)
+		{
+			returnValue = 0;
+			return make_int_sexp(1, &returnValue);
```

Comment:
This is one approach where we can create different numeric flags for various types of failure.